File: clojure/isbn-verifier/src/isbn_verifier.clj
(ns isbn-verifier)

(defn- valid-format? [isbn]
  (re-matches #"\d-?\d{3}-?\d{5}-?[\dX]" isbn))

(defn- char->digit [ch]
  (if (= ch \X)
    10
    (Character/digit ch 10)))

(defn- divisible-by? [divisor n]
  (zero? (mod n divisor)))

(defn- valid-checksum? [digits]
  (->> digits
       (remove #(= % \-))
       (map char->digit)
       (map * (range 10 0 -1))
       (apply +)
       (divisible-by? 11)))

(defn isbn? [isbn]
  (boolean (and (valid-format? isbn)
                (valid-checksum? isbn))))

File: haskell/isbn-verifier/src/IsbnVerifier.hs
module IsbnVerifier (isbn) where

import Data.Char (digitToInt, isDigit)

isbn :: String -> Bool
isbn xs = length digits == 10 && checksum `mod` 11 == 0
  where
    digits = isbnToDigits xs
    checksum = sum $ zipWith (*) digits [10, 9 .. 1]

isbnToDigits :: String -> [Int]
isbnToDigits [] = []
isbnToDigits ['X'] = [10]
isbnToDigits (x : xs)
  | isDigit x = digitToInt x : isbnToDigits xs
  | otherwise = isbnToDigits xs
